NitroxClient/GameLogic/Building.cs

```diff
@@ -1,213 +1,1326 @@
+
+        // When a gameObject is fully constructed and is to start deconstructing we also need to transfer the Id.
+        private NitroxId transferIdforConstructableBaseDeconstructing = null;
+
+        // Some objects like Ladders or Waterparks spawn surfaces on two floors. For this we need to know which surfaces are related.
+        private Dictionary<NitroxId, NitroxId> relationTopBottomIds = new Dictionary<NitroxId, NitroxId>();
+
+        // For the base objects themself as master objects of a base-complex we can't assign Ids to the ghosts, 
+        // 
+        private Dictionary<GameObject, NitroxId> baseGhostsIDCache = new Dictionary<GameObject, NitroxId>();
+
+
         public Building(IPacketSender packetSender, RotationMetadataFactory rotationMetadataFactory)
         {
             this.packetSender = packetSender;
             this.rotationMetadataFactory = rotationMetadataFactory;
         }
 
-        public void PlaceBasePiece(BaseGhost baseGhost, ConstructableBase constructableBase, Base targetBase, TechType techType, Quaternion quaternion)
+
+        // Section: Geometry Change Methods
+
+        // Start signal of a geometrychange for the viewable gameobjects
+        public void BaseGhost_Finish_Pre(BaseGhost baseGhost)
+        {
+#if TRACE && BUILDING
+            NitroxModel.Logger.Log.Debug("BaseGhost_Finish_Pre");
+#endif
+
+            baseGhostIsFinishing = true;
+        }
+
+        private string generateGameObjectTransformInfo(Transform transform)
+        {
+            return transform.name + " " + transform.parent.position.ToString() + " " + transform.GetSiblingIndex().ToString();
+        }
+
+        // Before the geometrychange we need to save the ids
+        public void Base_ClearGeometry_Pre(Base @base)
+        {
+            // Base_ClearGeometry_Pre is called twice. One time for the ghost layout and one time for the viewable layout. We are only interested in the viewable layout. 
+            {
+                idCacheForGeometryUpdate.Clear();
+
+                Transform[] cellObjects = (Transform[])@base.ReflectionGet("cellObjects");
+
+                if (cellObjects == null)
+                {
+                    return;
+                }
+
+                foreach (Transform cellObject in cellObjects)
+                {
+                    if (cellObject != null)
+                    {
+                        for (int i = 0; i < cellObject.childCount; i++)
+                        {
+                            Transform surface = cellObject.GetChild(i);
+
+                            if (surface != null && surface.gameObject != null)
+                            {
+
+#if TRACE && BUILDING
+                                string info = generateGameObjectTransformInfo(surface);
+                                NitroxModel.Logger.Log.Debug("Base_ClearGeometry_Pre - current clear for : " + info);
+#endif
+                                if (!surface.name.Contains("CorridorConnector"))
```

Comment:
CorridorConnectors have a special tech type and may be identified by these. But this will also enlarge this check, because there are different kinds of CorridorConnector for Hatches, Rooms, Moonpool, etc. The only common identification aspect i found was, that they all contain the string "CorridorConnector" in their names. Using a constant makes sense, i would put this as an internal enum in the BaseGeometry class, because it is just need there.